Avx jit

yekatkov opened the pull request.

NeoML-maintainer committed:
Build Framework (NeoML-master 1.0.45.0): Incrementing version number.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

NeoML-maintainer committed:
Build Framework (NeoML-master 1.0.45.0): Incrementing version number.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Merge remote-tracking branch 'upstream/master'

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Merge remote-tracking branch 'upstream/master'

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Merge remote-tracking branch 'upstream/master'

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Merge branch 'master' of https://github.com/yekatkov/neoml

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Merge remote-tracking branch 'upstream/master'

yekatkov committed:
Merge remote-tracking branch 'upstream/master'

yekatkov committed:
Merge branch 'master' of https://github.com/yekatkov/neoml

yekatkov committed:
Merge branch 'master' of https://github.com/yekatkov/neoml

yekatkov committed:
Init xbyak

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
MVP.
Added useJit parameter for testing.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
First success run.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Added single processing kernel.
Fixed some bugs.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Moved all register aliasing in calss members.
Move JIT initialization in BlobConvolution function.
Pass freeTerm pointer via Run parameter.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Added Jit debug.
Added xbyak headers.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Code refactoring

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Added FC 18 and 24.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
All work except FC 6

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
All tests pass.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Code refactoring.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Split code to several files.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Code refactoring

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Merge remote-tracking branch 'upstream/master' into avx_jit

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

# Conflicts:
#	NeoMathEngine/src/CPU/x86/avx/src/BlobConvolution.h

yekatkov committed:
Fix fetching parameters from the stack for Windows.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fix PR build

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fix build

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fix build

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fix build

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fix build

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fix build

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fix build

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fix build

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fix build

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Yet another way to fix build

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Added filter count 8

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Added FC 16.
Fixed bug in single processing FC 8.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Added FC 16.
Fixed bug in single processing FC 8.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Merge branch 'avx_jit' of https://github.com/yekatkov/neoml into avx_jit

yekatkov committed:
Fixed bug

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Merge branch 'avx_jit' of https://github.com/yekatkov/neoml into avx_jit

yekatkov committed:
Fixed bug on Windows

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

SAngeliuk commented on NeoMathEngine/src/CPU/x86/avx/src/xbyak/xbyak.h:
```diff
@@ -0,0 +1,2837 @@
+#pragma once
+#ifndef XBYAK_XBYAK_H_
```
We should write about xbyak here: https://github.com/neoml-lib/neoml#third-party

yekatkov committed:
Dummy commit

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Added vzeroupper instruction after JIT code.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Added FC 3.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Replaced non-VEX SIMD instruction with VEX ones.
Unroll FC24 batch processing and speed up FC24 single processing.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Unroll FC32

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fixed bug in batch FC32.
Unrolled single FC32.
Unrolled FC8.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fixed build

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Merge remote-tracking branch 'upstream/master' into avx_jit

yekatkov committed:
Added mention of xbyak to README.md

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Removed old code based on avx intrinsics.
Removed useJit argument (manual switching between jit and non-jit code).

In BlobConvolution_jit.inl fixed bugs:
1. Testing only low byte of regUseNarrowProcessing
2. Preserving xmm6-15 and regNumSteps.

Fixed potential read access violation in FC6, 18.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Code refactoring. A lot code was removed.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Code refactoring. Make result registers initialization common.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Added JIT test.
Fixed bug in getPixelOffsetSrcSteps().

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fixed build.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>

yekatkov committed:
Fixed Linux build.

Signed-off-by: Yury Katkov <ykatkov@yandex.ru>
